### PR TITLE
Azure Monitor cli commands for Managemnt & Data plane

### PR DESCRIPTION
**Pull request by vishrutshah**
This enables support for #1954 

Here is the complete set of available commands exposed from [azure-mgmt-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-mgmt-monitor) and [azure-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-monitor) python sdk. 

Here is the state of each Management command:

|  | Command  | Tested [Manual]  | Status  | Issue  | Category|
|---|---|---|---|---|---|
| log-profiles |   |   |   |   |
|  | create  |   |   | [Issue 973](https://github.com/Azure/azure-sdk-for-python/issues/973)  | Swagger |
|  | delete | :white_check_mark:  | :white_check_mark:  |   |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | list  | :white_check_mark:  | :white_check_mark:  |  |
|  | update  |   |   | [ Issue 973](https://github.com/Azure/azure-sdk-for-python/issues/973) |Swagger |
|---|---|---|---|---|
| alert-rule-incidents |   |   |   |   |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | list  | :white_check_mark:  | :white_check_mark:  |  |
|---|---|---|---|---|
| alert-rules  |   |   |   |   |
|  | create  |   |   | [Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971) [Issue 11](https://github.com/Azure/msrest-for-python/issues/11)  | AutoRest Python-Runtime |
|  | delete | :white_check_mark:  | :white_check_mark:  |   |
|  | show  |   |   | [Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971)  | AutoRest|
|  | list  |  |   |  [Issue 970](https://github.com/Azure/azure-sdk-for-python/issues/970)  | AutoRest|
|  | update  |   |   |[Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971) [Issue 11](https://github.com/Azure/msrest-for-python/issues/11) | AutoRest Python-Runtime |
|---|---|---|---|---|
| service-diagnostic-settings  |   |   |   |   |
|  | create  |   |   | [Issue 972](https://github.com/Azure/azure-sdk-for-python/issues/972)  | Swagger |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | update|   |   | [Issue 972](https://github.com/Azure/azure-sdk-for-python/issues/972)  | Swagger |
|---|---|---|---|---|
| autoscale-settings  |   |   |   | Known Issue in Service  | Monitor Service |
|  | create  | No  |   |  |
|  | delete |  No |   |   |
|  | show  | No |   |   |
|  | list  | No  |  |  | 
|  | update  |  No |   |  |

Here is the state of each Data Plane command:

|  | Command  | Tested [Manual]  | Status  | Issue  | Category |
|---|---|---|---|---|---|
| usage-metrics |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| event-categories |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| events |   |   |   |   | |
|  | list  |  |  | [Issue 969](https://github.com/Azure/azure-sdk-for-python/issues/969) | Swagger |
| tenant-events |   |   |   |   | |
|  | list  |  |  | [Issue 899](https://github.com/Azure/azure-rest-api-specs/issues/899) | Permission |
| metric-definitions |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| metrics |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |

**Note:** [azure-mgmt-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-mgmt-monitor) is not released as python package yet, once released we'll take our dependency on locally checked-in sdk before merge. 

@troydai @derekbekoe & @tjprescott when you get a chance, please have a first quick look. Thanks

### CODE REVIEW

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    def register_alias(self, argument_name, options_list):
+        register_cli_argument(self._commmand, argument_name, options_list=options_list)
+
+    def register_type(self, argument_name, options_list, input_type, arg_help):
+        register_cli_argument(self._commmand, argument_name,
+                              options_list=options_list, type=input_type, help=arg_help)
+
+    def expand(self, argument_name, model_type, group_name=None, patches=None):
+        # TODO:
+        # two privates symbols are imported here. they should be made public or this utility class
```
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: I think this would be a good time to move this utility class into azure.cli.core

**Comment by tjprescott** [comment]
You need to fix the linter issuer (add "# pylint: skip-file" to any auto-genned SDK) and also add a comment to the PR with the -h output for the most important commands.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from azure.cli.core.commands import register_cli_argument, cli_command
+from azure.cli.core.commands.parameters import ignore_type
+from azure.cli.core.commands.arm import cli_generic_update_command
+
+
+# MANAGEMENT CLIENT FACTORIES
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: These should go in a _client_factory.py file to be consistent with other packages.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 1 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: I disagree.  If done, it should be a separate PR to not make this review more difficult.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            type: group
+            short-summary: Commands to manage alert rule incidents.
+            """
+helps['monitor log-profiles'] = """
+            type: group
+            short-summary: Commands to manage log profiles.
+            """
+helps['monitor log-profiles update'] = """
+            type: command
+            short-summary: Update a log profile in Azure Monitoring REST API.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: No other command mentions a "REST API" in the CLI that I know of.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            type: group
+            short-summary: Commands to manage metric definitions.
+            """
+helps['monitor tenant-events'] = """
+            type: group
+            short-summary: Commands to manage tenant events.
+            """
+helps['monitor usage-metrics'] = """
+            type: group
+            short-summary: Commands to manage usage matrics.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: typo: metrics.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+
+helps['monitor'] = """
+            type: group
+            short-summary: Commands to manage Azure Monitor service.
+            """
+# MANAGEMENT COMMANDS HELPS
+helps['monitor alert-rules'] = """
+            type: group
+            short-summary: Commands to manage alert rules.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: It might make sense to clarify "manage Monitor alert rules" etc. for these commands since the terms are pretty generic and could be confused for other things.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import json
+from ._util import ParametersContext
+
+with ParametersContext(command='monitor alert-rules') as c:
+    c.register_alias('name', ('--azure-resource-name',))
+    c.register_alias('rule_name', ('--name', '-n'))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py]: The `--name` parameters here need a values completer within their scope (except for create commands).

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+with ParametersContext(command='monitor alert-rule-incidents') as c:
+    c.register_alias('incident_name', ('--name', '-n'))
+
+with ParametersContext(command='monitor log-profiles') as c:
+    c.register_alias('log_profile_name', ('--name', '-n'))
+
+with ParametersContext(command='monitor autoscale-settings') as c:
+    c.register_alias('name', ('--azure-resource-name',))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py]: `--azure-resource-name` is pretty verbose. Also, will this accept a name or id?  Should it?  If so, I would recommend `--resource` with help text that says "Name or ID of the resource to query." or something to that effect.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    c.register_alias('autoscale_setting_name', ('--name', '-n'))
+
+with ParametersContext(command='monitor alert-rules create') as c:
+    from azure.cli.command_modules.monitor.sdk.models.alert_rule_resource import AlertRuleResource
+
+    c.expand('parameters', AlertRuleResource)
+    c.register_type('condition', ('--condition',),
+                    json.loads,
+                    'JSON encoded condition configuration. Use @{file} to load from a file.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py]: I recommend you spell out the kwargs like we do in the traditional model. (i.e. type=json.loads, help='blahblah')

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+
+def get_complex_argument_processor(expanded_arguments, model_type):
+    """
+    Return a validator which will aggregate multiple arguments to one complex argument.
+    """
+    def _expension_valiator_impl(namespace):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py]: Typo: expansion

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import azure.cli.command_modules.monitor.help  # pylint: disable=unused-import
+
+__all__ = ['load_params', 'load_commands']
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py]: What's this line used for?
It shouldn't be needed.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/commands.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                    get_monitor_service_diagnostic_settings_operation,
+                    get_monitor_usage_metrics_operation, get_monitor_event_categories_operation,
+                    get_monitor_events_operation, get_monitor_tenant_events_operation,
+                    get_monitor_metric_definitions_operation, get_monitor_metrics_operation,
+                    ServiceGroup, create_service_adapter)
+
+# MANAGEMENT COMMANDS
+alert_rules_operations = create_service_adapter(
+    'azure.cli.command_modules.monitor.sdk.operations.alert_rules_operations',
+    'AlertRulesOperations')
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/commands.py]: This is another way of just doing `modpath.format(a, b, c)` like in the VM module (https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py#L19).

I don't really see this implementation as *simpler* than what's in the VM module as I now have to follow `create_service_adapter` to see what that does.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py (hunk 9 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py]: typo: validator

File: src/command_modules/azure-cli-monitor/setup.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    'Programming Language :: Python :: 3.4',
+    'Programming Language :: Python :: 3.5',
+    'Programming Language :: Python :: 3.6',
+    'License :: OSI Approved :: MIT License',
+]
+
+DEPENDENCIES = [
+    'azure-cli-core',
+    'azure-monitor==0.1.0'
+    # 'azure-mgmt-monitor==0.1.0'
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/setup.py]: Please remove if it's not used.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 1 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: Makes sense. Then I'd prefer it to refactor once this PR gets in.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py (hunk 3 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: Ya that's true..that came from swagger comment...I'll remove it 👍

File: src/command_modules/azure-cli-monitor/setup.py (hunk 12 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/setup.py]: @derekbekoe I was thinking that we'll not merge the PR in as it has sdk checked-in. So I was planning to wait until azure-mgmt-monitor releases and then removing sdks and the merging PR. 

Let me know if there is way where we still allow merging PRs if sdks are checked-in? If so i'll definitely remove it.

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor -h

Group
    az monitor: Commands to manage Azure Monitor service.

Subgroups:
    alert-rule-incidents       : Commands to manage alert rule incidents.
    alert-rules                : Commands to manage alert rules.
    autoscale-settings         : Commands to manage autoscale settings.
    event-categories           : Commands to manage event categories.
    events                     : Commands to manage events.
    log-profiles               : Commands to manage log profiles.
    metric-definitions         : Commands to manage metric definitions.
    metrics                    : Commands to manage metrics.
    service-diagnostic-settings: Commands to manage service diagnostic settings.
    tenant-events              : Commands to manage tenant events.
    usage-metrics              : Commands to manage usage matrics.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor alert-rule-incidents -h

Group
    az monitor alert-rule-incidents: Commands to manage alert rule incidents.

Commands:
    list: Gets a list of incidents associated to an alert rule.
    show: Gets an incident associated to an alert rule.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor alert-rules -h

Group
    az monitor alert-rules: Commands to manage alert rules.

Commands:
    create: Creates or updates an alert rule.
    delete: Deletes an alert rule.
    list  : List the alert rules within a resource group.
    show  : Gets an alert rule.
    update: Updates an alert rule.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor autoscale-settings -h

Group
    az monitor autoscale-settings: Commands to manage autoscale settings.

Commands:
    create: Creates or updates an autoscale setting.
    delete: Deletes and autoscale setting.
    list  : Lists the autoscale settings for a resource group.
    show  : Gets an autoscale setting.
    update: Updates an autoscale setting.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@visshamac azure-cli-monitor (monitor-cli) $ az monitor event-categories -h

Group
    az monitor event-categories: Commands to manage event categories.

Commands:
    list: Get the list of available event categories supported in the Activity Log Service.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor events -h

Group
    az monitor events: Commands to manage events.

Commands:
    list: Provides the list of events.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@visshamac azure-cli-monitor (monitor-cli) $ az monitor log-profiles -h

Group
    az monitor log-profiles: Commands to manage log profiles.

Commands:
    create: Create or update a log profile in Azure Monitoring REST API.
    delete: Deletes the log profile.
    list  : List the log profiles.
    show  : Gets the log profile.
    update: Update a log profile in Azure Monitoring REST API.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor metric-definitions -h

Group
    az monitor metric-definitions: Commands to manage metric definitions.

Commands:
    list: Lists the metric definitions for the resource.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor metrics -h

Group
    az monitor metrics: Commands to manage metrics.

Commands:
    list: Lists the metric values for a resource.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor service-diagnostic-settings -h

Group
    az monitor service-diagnostic-settings: Commands to manage service diagnostic settings.

Commands:
    create: Create or update new diagnostic settings for the specified resource.
    show  : Gets the active diagnostic settings for the specified resource.
    update: Update diagnostic settings for the specified resource.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor tenant-events -h

Group
    az monitor tenant-events: Commands to manage tenant events.

Commands:
    list: Get the Activity Logs for the Tenant.
```

**Comment by vishrutshah** [comment]
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor usage-metrics -h

Group
    az monitor usage-metrics: Commands to manage usage matrics.

Commands:
    list: The List operation lists the usage metrics for the resource.
```

**Comment by vishrutshah** [comment]
@tjprescott & @derekbekoe I have just put some more comments on showing the documentation around commands for the reference. I haven't gone through all your comments but as I go I'll keep fixing them. This is just in case you want to see the commands itself 💯 

I'll ping you guys once i fix all the review comments :)

**Comment by tjprescott** [comment]
The help text for several of the list commands illustrates a major problem. 

First, the text from the SDK is crazily verbose, and we should replace it, but second, we should look at the equivalent Xplat CLI commands to see what convenience arguments they use to build up a filter. We should **not** simply expose the `--filter` parameter and tell people to go learn OData. Even if we do expose `--filter`, we should allow the building of a filter with argument.  See the `az resource list` command for examples of this.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py (hunk 10 above)
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py]: It's just a formality though it is not technically needed.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def group(self, group_name):
+        return CommandGroup(self._scope, group_name, self._factory, self._service_adapter)
+
+
+class CommandGroup(object):
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: This is duplicate with the sql command module. Consider move it up to the `azure-cli-core` for reuse.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+    def expand(self, argument_name, model_type, group_name=None, patches=None):
+        # TODO:
+        # two privates symbols are imported here. they should be made public or this utility class
+        # should be moved into azure.cli.core
+        from azure.cli.core.commands import _cli_extra_argument_registry
+        from azure.cli.core.commands._introspection import \
+            (extract_args_from_signature, _option_descriptions)
+
+        from .validators import get_complex_argument_processor
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: Prefer absolute path

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/sdk/__init__.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
+# coding=utf-8
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/sdk/__init__.py]: are these coming from SDK? shall we commit this?

File: src/command_modules/azure-cli-monitor/setup.py (hunk 12 above)
Comment (troydai) [inline src/command_modules/azure-cli-monitor/setup.py]: No. We should NOT. No matter If azure-mgmt-monitor is released we should not include their source code in our code base.

Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/setup.py]: Okay didn't know you weren't planning on merging this right now.
I've created and added a Do Not Merge label to this PR then.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py (hunk 5 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: Updated: Commands to manage alerts assigned to Azure resources

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py (hunk 3 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: Updated: Update a log profile assigned to Azure subscription

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py (hunk 4 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py]: Fixed

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py (hunk 7 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py]: I haven't done tests on autoscale, so i am not sure on whether it can accept ids. So I'll keep this pending until then :)

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py (hunk 8 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py]: Cool

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py (hunk 9 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py]: fixed

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py (hunk 10 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py]: removed

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/commands.py (hunk 11 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/commands.py]: I see but as I am following the flow/structure from `azure-cli-sql` (as I have understood that flow as of now :)), I'd prefer to keep it same. What's your thought on that?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 13 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: Yes, we should. I'd prefer it to do that refactoring once this PR completes so I don't confuse you guys more on actual PR. Does that seem good?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/sdk/__init__.py (hunk 15 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/sdk/__init__.py]: Nope. Will not merge in until azure-mgmt-monitor is released. PR https://github.com/Azure/azure-sdk-for-python/pull/974

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 14 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: Updated to azure.cli.command_modules.monitor.validators

File: src/command_modules/azure-cli-monitor/setup.py (hunk 12 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-monitor/setup.py]: We have allowed services (KeyVault, IoT, etc.) to merge PRs with embedded clients. DocumentDB was about to earlier this week but the SDK was released so he went ahead and did the import changes.

If something holds up this PR from being merged, it really doesn't need to be the desire to avoid (temporarily) having the SDK code in the module.

**Comment by vishrutshah** [comment]
@troydai @tjprescott @derekbekoe 

I've removed SDK code from the cli and took dependencies on `azure-mgmt-monitor` [version 0.1.0](https://pypi.python.org/pypi/azure-mgmt-monitor/0.1.0) & `azure-monitor` [version 0.2.0](https://pypi.python.org/pypi/azure-monitor/0.2.0) sdks. 

I've also addressed general code review comments except:

1. As I am following design of `azure-cli-sql`, we have some design that do not match with the other modules. Example: I do not have _client_factory.py, I follow ParametersContext to register commands, expand arguments and aliasing etc
2. I did duplicate common code between `azure-cli-sql` & `azure-monitor-cli` that will be moved and shared as separate PR
3. I did not provide very convenience methods around --filter parameters yet. I'd like to provide those convenience as separate PR
4. I've tested commands manually and currently writing automation tests around it. I can add them here or I can do that separate to make it easier for me and reviewer :)

Let me know what you guys think about new changes and how do we want to proceed.

Thanks!

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+def get_monitor_log_profiles_operation(kwargs):
+    return get_monitor_management_client(kwargs).log_profiles
+
+
+# DATA CLIENT FACTORIES
+def get_monitor_client(_):
+    from azure.monitor import MonitorClient
+    from azure.cli.core.commands.client_factory import get_mgmt_service_client
+    return get_mgmt_service_client(MonitorClient)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: How does this work if you're passing in the data-plane client yet there's a call to `get_mgmt_service_client`?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 2 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: Still think the client factories should be in _client_factory.py file.

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py (hunk 16 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: strange...looks like i missed something here..Thanks for catching..Let me verify

Comment (vishrutshah) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py]: It is indeed odd but it works :)

```
(env) vishrut@mac azure-cli (monitor-cli) $ az monitor event-categories list
[
  {
    "localizedValue": "Administrative",
    "value": "Administrative"
  },
  {
    "localizedValue": "Security",
    "value": "Security"
  },
  {
    "localizedValue": "Service Health",
    "value": "ServiceHealth"
  },
  {
    "localizedValue": "Alert",
    "value": "Alert"
  },
  {
    "localizedValue": "Recommendation",
    "value": "Recommendation"
  },
  {
    "localizedValue": "Policy",
    "value": "Policy"
  }
]
(env) vishrut@amac azure-cli (monitor-cli)
```

The reason I believe that azure-monitor & azure-mgmt-monitor sdks are generated from `Azure.Python` generator, hence the signature of client initialization is the same in both management plane and data plane sdks. So I did not ran into the any problem.

**Comment by tjprescott** [comment]
I would not recommend merging this before the release while it exposes raw "--filter" with no helper arguments.

**Comment by vishrutshah** [comment]
Implement filter for `metrics-definitions list` command PR: https://github.com/vishrutshah/azure-cli/pull/1. Please take a look when you get a chance. Thanks!

**Comment by vishrutshah** [comment]
Implement filter for `metrics list` command PR: https://github.com/vishrutshah/azure-cli/pull/2. Please take a look when you get a chance. Thanks!

**Comment by tjprescott** [comment]
@vishrutshah which commands remain that have filter building arguments in Xplat but not in this PR?

**Comment by vishrutshah** [comment]
Implement filter for `activity-logs list` & `tenant-activity-logs list` command PR: https://github.com/vishrutshah/azure-cli/pull/3. Please take a look when you get a chance. Thanks! 

@tjprescott With this PR merged in, we have following 2 commands from _management plane_ left  to be simplified:

```
az monitor alert-rules list
```
```
az monitor autoscale-settings list
```

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py [hunk 17]
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
...
+    :param str resource_group: The resource group
+    :param str resource_id: The identifier of the resource
+    :param str resource_provider: The resource provider
+    :param str start_time: The start time of the query. In ISO format with explicit indication of
+                           timezone: 1970-01-01T00:00:00Z, 1970-01-01T00:00:00-0500
+    :param str end_time: The end time of the query. In ISO format with explicit indication of
+                         timezone: 1970-01-01T00:00:00Z, 1970-01-01T00:00:00-0500
+    :param str caller: The caller to look for when querying
+    :param str status: The status value to query (ex: Failed)
+    :param str max_events: The maximum number of records to be returned by the command
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py]: why is it `str` not `int`?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py [hunk 18]
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
...
+        raise CLIError("usage error: [--correlation-id ID | --resource-group NAME | "
+                       "--resource-id ID | --resource-provider PROVIDER]")
+
+    odata_filters = _build_activity_logs_odata_filter(correlation_id, resource_group,
+                                                      resource_id, resource_provider,
+                                                      start_time, end_time,
+                                                      caller, status)
+
+    if max_events:
+        max_events = int(max_events)
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py]: can't `argparser` or validator guarantee it's integer?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py [hunk 19]
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
...
+    :param str start_time: The start time of the query. In ISO format with explicit indication of
+                           timezone: 1970-01-01T00:00:00Z, 1970-01-01T00:00:00-0500
+    :param str end_time: The end time of the query. In ISO format with explicit indication of
+                         timezone: 1970-01-01T00:00:00Z, 1970-01-01T00:00:00-0500
+    :param str caller: The caller to look for when querying
+    :param str status: The status value to query (ex: Failed)
+    '''
+    formatter = "eventTimestamp ge {} and eventTimestamp le {}"
+    odata_filters = _validate_time_range_and_add_defaults(start_time, end_time,
+                                                          formatter=formatter)
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py]: should time validating be done in `validator`?

File: src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py [hunk 20]
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
...
+                                            caller, 'caller')
+    if status:
+        odata_filters = _build_odata_filter(odata_filters, 'status',
+                                            status, 'status')
+
+    return odata_filters
+
+
+def _activity_logs_select_filter_builder(events=None):
+    '''Build up select filter string from events
```
Comment (troydai) [inline src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py]: use triple double quote for docstring.

**Comment by tjprescott** [comment]
As we discussed Friday, we should close this PR until it is ready to be merged and then either reopen it or open a new one.

**Comment by vishrutshah** [comment]
Closing this PR and sending, new one for the review for partner team.

**Comment by vishrutshah** [comment]
New PR @ https://github.com/Azure/azure-cli/pull/2534
